pkg/ipmi/ipmi.go

```diff
@@ -0,0 +1,176 @@
+// Copyright 2019 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package ipmi implements functions to communicate with
+// the OpenIPMI driver interface.
+package ipmi
+
+import (
+	"os"
+	"syscall"
+	"unsafe"
+)
+
+const (
+	_IOC_WRITE = 0x1
+	_IOC_READ  = 0x2
+
+	_IOC_NRBITS   = 8
+	_IOC_TYPEBITS = 8
+	_IOC_SIZEBITS = 14
+	_IOC_NRSHIFT  = 0
+
+	_IOC_TYPESHIFT = _IOC_NRSHIFT + _IOC_NRBITS
+	_IOC_SIZESHIFT = _IOC_TYPESHIFT + _IOC_TYPEBITS
+	_IOC_DIRSHIFT  = _IOC_SIZESHIFT + _IOC_SIZEBITS
+
+	_IPMI_BMC_CHANNEL                = 0xf
+	_IPMI_BUF_SIZE                   = 1024
+	_IPMI_IOC_MAGIC                  = 'i'
+	_IPMI_NETFN_APP                  = 0x6
+	_IPMI_OPENIPMI_READ_TIMEOUT      = 15
+	_IPMI_SYSTEM_INTERFACE_ADDR_TYPE = 0x0c
+
+	_BMC_SET_WATCHDOG_TIMER = 0x24
+	_BMC_GET_WATCHDOG_TIMER = 0x25
+
+	_IPM_WATCHDOG_NO_ACTION    = 0x00
+	_IPM_WATCHDOG_SMS_OS       = 0x04
+	_IPM_WATCHDOG_CLEAR_SMS_OS = 0x10
+)
+
+var (
+	_IPMICTL_RECEIVE_MSG  = iowr(_IPMI_IOC_MAGIC, 12, int(unsafe.Sizeof(recv{})))
+	_IPMICTL_SEND_COMMAND = ior(_IPMI_IOC_MAGIC, 13, int(unsafe.Sizeof(req{})))
+)
+
+type IPMI struct {
+	*os.File
+}
+
+type msg struct {
+	netfn   byte
+	cmd     byte
+	dataLen uint16
+	data    unsafe.Pointer
+}
+
+type req struct {
+	addr    *systemInterfaceAddr
+	addrLen uint32
+	msgid   int64
```

Comment:
optional: do you want to annotate these lines with nolint? https://github.com/golangci/golangci-lint#nolint

```go
msgid int64 //nolint:structcheck
```